### PR TITLE
Volttron7 simulation

### LINKED ISSUE #2423: Integration of external simulation platforms with VOLTTRON
Create a simulation integration framework that allows one to integrate VOLTTRON with different type of simulation/co-simulation platforms like HELICS, GridAPPSD, EnergyPlus easily.

### PR DESCRIPTION
**Pull request by shwethanidd**
# Description

This pull request contains simulation integration feature that allows VOLTTRON to integrate with external simulation environments using standardized simulation integration framework (similar to driver or historian framework.

The first pull request contains the

    1. Design document describing baseSimulation Integration framework
    2. Integration with HELICS co-sim
    3. Integration with GridAPPSD simulation
    4. Integration with EnergyPlus simulation

Fixes #2423

### CODE REVIEW

**Comment by craig8** [comment]
I think we should mention in the rst doc that we have implemented the following agents in the examples folder of the repository...

probably examples/integrations b/c there are so many examples.  Thoughts?

File: examples/EnergyPlusAgent/README.md [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+
+```
+wget https://github.com/NREL/EnergyPlus/releases/download/v8.5.0/EnergyPlus-8.5.0-c87e61b44b-Linux-x86_64.sh
+chmod +x EnergyPlus-8.5.0-c87e61b44b-Linux-x86_64.sh
+sudo ./EnergyPlus-8.5.0-c87e61b44b-Linux-x86_64.sh 
+```
+
+## EnergyPlus Agent Configuration
+
+1. Copy 'bcvtb' and 'eplus' folders into root of VOLTTRON source directories. 'eplus' should contain EnergyPlus Input Data File (IDF). IDF file is an ASCII file containing the data describing the building and HVAC system to be simulated.
```
Comment (craig8) [inline examples/EnergyPlusAgent/README.md:27]: Where do I get these folders from?

**Comment by craig8** [comment]
@shwethanidd can you do a merge from develop into the branch

**Comment by shwethanidd** [comment]
Yes I did yesterday.

**Comment by craig8** [comment]
Great!

Get Outlook for Android<https://aka.ms/ghei36>
________________________________

File: examples/EnergyPlusAgent/requirements.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+
```
Comment (schandrika) [inline examples/EnergyPlusAgent/requirements.txt:1]: is this file required?

File: examples/GridAPPS-D/README.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
+# GridAPPSD Simulation Example Agent
+
+This is an example agent that demonstrates how to integrate with GridAPPSD platform, 
+run power system simulations and send/receive messages back and forth between VOLTTRON and
+GridAPPSD environment. 
+
+## GridAPPSD installation
+For installing setup in Ubuntu based systems, follow the steps described in 
+https://gridappsd.readthedocs.io/en/master/installing_gridappsd/index.html
+
```
Comment (schandrika) [inline examples/GridAPPS-D/README.md:10]: Provide link to simulation framework docs

**Review by schandrika — CHANGES_REQUESTED** [review]

File: docs/source/specifications/simulation_integration.rst [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+3. Publish outputs from simulation to VOLTTRON
+4. Subscribe to topics from VOLTTRON
+5. Send inputs to simulation
+6. Advance simulation time step
+7. Pause simulation
+8. Resume simulation
+9. Stop simulation
+
+Currently, VOLTTRON has individual implementations for integrating with
+many of the above simulation platforms. Instead, in this specification we
```
[the remark this answers is not in the document]
Comment (shwethanidd) [reply docs/source/specifications/simulation_integration.rst]: Provided links. They will still be available to users (especially existing energyplus agent in VOLTTRON-GS repo)

File: examples/EnergyPlusAgent/README.md (hunk 1 above)
Comment (shwethanidd) [reply examples/EnergyPlusAgent/README.md:28]: I have added link to 'eplus'. Waiting on Robert to give me link to 'bcvtb'.

**Review by shwethanidd — COMMENTED** [review]
Fixed review comments

**Review by sgilbride — CHANGES_REQUESTED** [review]

File: docs/source/core_services/integrations/Simulation-Integration.rst [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+    for pub in self.publications:
+        key = pub['sim_topic']
+        value = 67.90
+        self.helics_sim.publish_to_simulation(key, value)
+
+Advance the simulation
+----------------------
+With some simulation platforms such as HELICS, the federate can make explicit time request to advance in time by certain
+number of time steps. There will be a global time keeper (in this case HELICS broker) which will be responsible of maintaining
```
[the remark this answers is not in the document]
Comment (shwethanidd) [reply docs/source/core_services/integrations/Simulation-Integration.rst:119]: done

File: docs/source/core_services/integrations/Simulation-Integration.rst [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+time within the simulation. In the time request mode, each federate has to request for time advancement after it has
+completed it's work. The global time keeper grants the lowest time among the all time requests. All the federates receive
```
[the remark this answers is not in the document]
Comment (shwethanidd) [reply docs/source/core_services/integrations/Simulation-Integration.rst:121]: done

File: docs/source/core_services/integrations/Simulation-Integration.rst [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+capability of pause/resume simulation, so calling resume_simulation() API will result in no operation.
+
+.. code-block:: python
+
+    self.helics_sim.resume_simulation()
+
+Stop the simulation
+-------------------
+The agent can stop the simulation at any point of point. In the case of HELICSSimIntegration object, it will disconnect
+the federate from the HELICS core and close the library. Generally, it is a good practise to call the stop_simulation API
```
[the remark this answers is not in the document]
Comment (shwethanidd) [reply docs/source/core_services/integrations/Simulation-Integration.rst:154]: done

File: examples/EnergyPlusAgent/README.md [hunk 8]
```diff
@@ -0,0 +1,255 @@
+# EnergyPlus Agent
+
+This is an example agent that demonstrates how to integrate with EnergyPlus simulation, run building model simulations 
+with EnergyPlus, send/receive messages back and forth between VOLTTRON and
+EnergyPlus simulation. For more information about EnergyPlus, please refer to https://www.energyplus.net/sites/default/files/docs/site_v8.3.0/GettingStarted/GettingStarted/index.html  
+
```
[the remark this answers is not in the document]
Comment (shwethanidd) [reply examples/EnergyPlusAgent/README.md:7]: done

File: examples/EnergyPlusAgent/README.md [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+## EnergyPlus Agent Configuration
+
+1. Copy 'bcvtb' and 'eplus' folders into root of VOLTTRON source directories. 'eplus' should contain EnergyPlus Input Data File (IDF). IDF file is an ASCII file containing the data describing the building and HVAC system to be simulated.
+
+'eplus' is available at https://github.com/VOLTTRON/volttron-GS/tree/master/eplus
+
+2. You can specify the configuration in either json or yaml format.  The json format is specified
+below. 
+
+```` yaml
```
[the remark this answers is not in the document]
Comment (shwethanidd) [reply examples/EnergyPlusAgent/README.md:35]: done

File: examples/GridAPPS-D/README.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
+# GridAPPSD Simulation Example Agent
```
[the remark this answers is not in the document]
Comment (shwethanidd) [reply examples/GridAPPS-D/README.md:1]: done

**Review by shwethanidd — COMMENTED** [review]
Incorporated all review comments
